Initial stuff

cryogenian opened the pull request:
Printer, search interpreter and `JArray` example

@garyb Could you please take a look?

cryogenian committed:
constructors

cryogenian committed:
wip

cryogenian committed:
lenses

cryogenian committed:
more lenses

cryogenian committed:
...

cryogenian committed:
wip examples

cryogenian committed:
search example

cryogenian committed:
joggling packages wip

cryogenian committed:
tests...

cryogenian committed:
only tests for search interpreter is left

cryogenian committed:
done

cryogenian committed:
readme

garyb commented on src/SqlSquare/AST.purs:
```diff
@@ -1,225 +1,125 @@
+
+import Matryoshka (class Recursive, Algebra, cata, transParaT, project)
+
+import Debug.Trace as DT
```
Debug.Trace :smile:

garyb commented on src/SqlSquare/AST.purs:
```diff
@@ -1,225 +1,125 @@
-      , orderBy ∷ Maybe (OrderBy a)
-      }
+  | Select (SelectR a)
+  | Parens a
```
Do you think we could use EJson rather than having constructors for all the literals in here? I don't think there are any EJson literals that don't have a corresponding SQL^2 representation. SQL^2 will need an additional set literal though.

If there's some problem with that I'm not thinking of, I have another related suggestion - although this is kinda a personal-preference thing - there might be some value in separating literals out into their own ADT, and just having a `Literal (LiteralF a)` constructor in `SqlF`. It simplified some stuff when I applied it in PS anyway.

jdegoes replied:
Yes, using EJson is a good idea. Technically you could use `Coproduct SqlF EJsonF` where `SqlF` is the part of SQL<sup>2</sup> that is not `EJson`.

Remember in SQL<sup>2</sup> you can do things like this:

```
SELECT {d.dynamic_field : [d.another_one]} FROM `/foo/` AS d
```

i.e. "literals" can recursively contain arbitrary SQL<sup>2</sup> / EJSON.

garyb commented on src/SqlSquare/AST.purs:
```diff
@@ -267,19 +167,21 @@ instance functorAST ∷ Functor SqlF where
     ArrayLiteral lst →
       ArrayLiteral $ map f lst
+    Parens t →
+      Parens $ f t
```
Is this not a derivable functor?

cryogenian replied:
Nope, there is `Tuple a a`

garyb commented on src/SqlSquare/JoinType.purs:
```diff
@@ -0,0 +1,19 @@
+  LeftJoin → "left join"
+  RightJoin → "right join"
+  FullJoin → "full join"
+  InnerJoin → "inner join"
```
This is a general comment, but here seems like a good a place as any... should we render the SQL-syntax-stuff as all uppercase perhaps? That seems more conventional to me; and I tend to write it that way when doing so by hand as it makes it easier to spot the parts that are identifiers, etc.

jdegoes replied:
👍

garyb commented on src/SqlSquare/Lenses.purs:
```diff
@@ -0,0 +1,232 @@
+import SqlSquare.Utils (type (×), (∘), (⋙))
+
+_Newtype ∷ ∀ n t. Newtype n t ⇒ Iso' n t
+_Newtype = iso unwrap wrap
```
There's already an iso for this: https://pursuit.purescript.org/packages/purescript-profunctor-lenses/2.6.0/docs/Data.Lens.Iso.Newtype#v:_Newtype

garyb reviewed (changes requested):
A couple of comments, I'll stop for now and then resume again once we've decided whether to do the EJson stuff, etc.

cryogenian committed:
Literal, need manually implement Ord1 and Eq1

garyb commented on src/SqlSquare/Constructors.purs:
```diff
@@ -2,60 +2,67 @@ module SqlSquare.Constructors where
 import SqlSquare.AST  (SqlF(..), Relation, GroupBy(..), OrderBy, BinaryOperator, UnaryOperator, (∘), SelectR, Case(..), Projection(..))
 
-vari ∷ ∀ t. Corecursive t SqlF ⇒ String → t
+vari ∷ ∀ t. Corecursive t (SqlF EJsonF) ⇒ String → t
```
I think we can leave `EJsonF` polymorphic for a bunch of these...

cryogenian committed:
using EJsonF

cryogenian committed:
updated ejson dep

cryogenian committed:
capitalized

cryogenian committed:
capitalize tests, polymorphic l

cryogenian commented:
+ Removed `Debug.Trace`
+ AST has polymorphic `Literal l`, it's `EJsonF` here 
+ Removed `_Newtype`
+ Capitalized keywords
+ Made smart constructors polymorphic

cryogenian commented:
Ah! Yep, added `Foldable` and `Traversable` instances. Actually `Search` example could be rewritten using `cataM` for checking fields/labels but it's really tedious :)

cryogenian committed:
updated to the latest purescript-ejson, added date|time constructors|lenses

cryogenian commented:
+ Updated to `ejson` with proper date|time 
+ Added constructors and lenses for date|time literal

@garyb is this good to go?

garyb commented on src/SqlSquare/OrderBy.purs:
```diff
@@ -0,0 +1,32 @@
+  foldr f a (OrderBy xs) = F.foldr (flip (F.foldr f)) a xs
+instance traversableOrderBy ∷ T.Traversable OrderBy where
+  traverse f (OrderBy xs) = map OrderBy $ T.traverse (T.traverse f) xs
+  sequence = T.sequenceDefault
```
I think `derive newtype instance` should work for `Foldable` / `Traversable` here?

garyb commented on src/SqlSquare/Lenses.purs:
```diff
@@ -0,0 +1,318 @@
+
+_Splice
+  ∷ ∀ t
+  . (Recursive t (S.SqlF EJ.EJsonF), Corecursive t (S.SqlF EJ.EJsonF))
```
This is a pretty minor thing, but we could make the lenses that aren't literal-specific polymorphic in `SqlF l` too perhaps. I wouldn't hold up this PR for that though!

garyb reviewed (approved):
It's going to be great to be able to replace our string munging with this!

cryogenian committed:
lenses

cryogenian commented:
I've made lenses more polymorphic, but `OrderBy` foldable instance isn't derivable :(

garyb commented:
Oh well! Just thought it was worth a try.
